Refresh.GameServer/Database/GameDatabaseContext.Activity.cs

```diff
@@ -28,6 +28,24 @@ public partial class GameDatabaseContext // Activity
             .Where(e => e.Timestamp < timestamp && e.Timestamp >= endTimestamp)
             .AsEnumerable();
 
+        if (user != null)
+        {
+            List<ObjectId?> favouriteUsers = user.UsersFavourited.AsEnumerable().Select(f => (ObjectId?)f.UserToFavourite.UserId).ToList();
+            List<ObjectId?>? userFriends = friendService?.GetUsersFriends(user, this)?.Select(u => (ObjectId?)u.UserId).ToList();
+
+            query = query.Where(e =>
+                e.User.UserId == user.UserId ||
+                e.StoredObjectId == user.UserId ||
+                favouriteUsers.Contains(e.User.UserId) ||
+                favouriteUsers.Contains(e.StoredObjectId) ||
+                (userFriends?.Contains(e.User.UserId) ?? false) ||
+                (userFriends?.Contains(e.StoredObjectId) ?? false) ||
+                this.GetLevelById(e.StoredSequentialId ?? int.MaxValue)?.Publisher?.UserId == user.UserId ||
+                e.EventType == EventType.Level_TeamPick ||
+                e.EventType == EventType.User_FirstLogin
```

Comment:
Some of this should *definitely* be moved to the main `IQueryable` call in at the top of the function if possible. This is going to be a performance nightmare.

Reply:
> Some of this should _definitely_ be moved to the main `IQueryable` call in at the top of the function if possible. This is going to be a performance nightmare.

Cant be on IQueryable, indirect accesses (like `e.User.UserId`) are unsupported by Realm

Reply:
Some of these queries, yes, but not all. `e.StoredObjectId`, `e.EventType` checks for example can be moved earlier.

Reply:
If its a big issue, then I recommend extracting the filtering logic to a new function, and then split this function in to 3 separate ones for levels, users, and global queries. I think that just makes more sense in general, too.

Reply:
Im not sure how to write this logic in a fashion where i *can* split it up into multiple `.Where` calls. This filters to where the event type is *any* of these many possible things. I'd have to do two `.Where` calls (both on the raw query, cant be chained), then do `.Union` to create a union of the two queries, which im not sure how much more performant that would be, given now im enumerating three times (one per `.Where`, one for the `.Union`).

Reply:
Right, makes sense. I'm vehemently against how long that takes (250ms is frankly absurd especially given the power of your workstation) but if there's no better way to do it then I guess it'll have to work.

This is something that we'll have to consider when we get PostgreSQL support in a functional state. This makes me wonder if we can do raw Realm queries somehow.

That said, we can still the refactoring I suggested. Bunching together logic for what's ultimately 3 different types of result is just a disaster.

Reply:
Yeah performance im really not happy with (in release mode even after the JIT warms up its ~100ms). Just to be sure i understand what you mean with the refactor, do you mean having separate functions for "get filtered global recent activity", "get filtered recent activity scoped to this user", "get filtered global recent activity for a level"?

Reply:
Yep.